Extended unit tests for string hashing

charliescheer opened the pull request:
### Fix
This PR responds to issue #1053 and adds additional unit tests to confirm the hashing done by the string method byEncodingAsTagHash

Some tests for this method already existed, so in this PR I have extended the test "testByEncodingTagAsHashAllowsUsToProperlyCompareStringsThatEvaluateAsNotEqualOtherwise" to include some additional use cases, the cases specifically called for in #1053, and added a test to confirm the encoding of the special character set defined in the issue.

The new test that I created may be repeating the test before it, so may be unnecessary but take a look, I thought specifically checking the special characters would prove useful.  There may also be a more efficient way to get a similar character set than writing out the encoded hash as I did.

### Test
1. in Xcode go to NSStringSimplenoteTests.swift
2. Run the unit tests for the project
3. All tests should pass

### Review

> Only one developer is required to review these changes, but anyone can perform the review.

### Release
> These changes do not require release notes.

charliescheer committed:
Added test verifying hash of special characters with byEncodingAsTagHash

charliescheer committed:
Added additional asserts to verify hash values that evaluate equal

eshurakov commented on SimplenoteTests/NSStringSimplenoteTests.swift:
```diff
@@ -28,17 +38,29 @@ class NSStringSimplenoteTests: XCTestCase {
         let sampleA = NSString(stringLiteral: "\u{0073}\u{0323}\u{0307}")
         let sampleB = NSString(stringLiteral: "\u{0073}\u{0307}\u{0323}")
         let sampleC = NSString(stringLiteral: "\u{1E69}")
+        let sampleD = NSString(stringLiteral: "\u{0065}\u{0301}")
+        let sampleE = NSString(stringLiteral: "\u{00E9}")
```
I think it's better to separate this test in two, because it is testing 2 sets of input data which are not related to each other. 
We can extract a function that would accept an array of strings and test them according to existing logic: check that strings in array are not equal and check that hashes are equal. After that we can call this function with different sets of data from our test.

Example pseudo code:
```
func testNonEqualStringsHaveTheSameHash(_ samples: [String]) { ... }
func test...() {
    testNonEqualStringsHaveTheSameHash(["\u{0073}\u{0323}\u{0307}", "\u{0073}\u{0307}\u{0323}", ""\u{1E69}""])
    testNonEqualStringsHaveTheSameHash(["\u{0065}\u{0301}", "\u{00E9}"])
}
```

charliescheer replied:
That is a good idea for separating the different values that should test the same.  I updated the tests, can you check it out again when you get a chance and see what you think?

I'm not sure it is the clearest on what it is testing, if you have any suggestions for improving the readability that would be great!

eshurakov replied:
Looks good! But we can do better! :-)

First we can make `func testNonEqualStringsCreateSameHash(_ samples: [NSString])` private, because it's our internal helper method.

To make it more readable, we can combine samples in array like this:
```
let differentStringsOneHashSamples = [
    ["\u{0073}\u{0323}\u{0307}", "\u{0073}\u{0307}\u{0323}", "\u{1E69}"], // ṩ
    ["\u{0065}\u{0301}", "\u{00E9}"] // é
]
```

And in the test we can walk through our samples and call helper method `testNonEqualStringsCreateSameHash`. I think we can also rename the original test `testByEncodingTagAsHashAllowsUsToProperlyCompareStringsThatEvaluateAsNotEqualOtherwise` to `testNonEqualStringsCreateSameHash` but without parameters.

In `differentStringsOneHashSamples` I used swift strings instead of NSString, I think we don't really need to stick to NSString for most of the test, so it's fine to use Swift strings. The only "issue" is that comparing our samples using swift strings would return `true` as it seems that swift strings compare "characters" and not  underlying representation of the string. To solve this we can temporarily cast our strings to NSString: `XCTAssertFalse((samples[count] as NSString).equal(to: samples[testCount]))`

Another thing is that in original test we were comparing every sample to every sample, but I don't think it's needed. I think we can compare the first sample to all other samples and get the same coverage. If A == B and A == C, then B == C. This way we will need to walk the samples array only once.

p.s. you can use for loop instead of while + separate var: `for i in 0..<samples.count`

eshurakov reviewed (changes requested):
Hey @charliescheer! 
Sending you a suggestion that I think will make this test better! Let me know what do you think.

charliescheer committed:
Added helper function to test if  non equal strings hash the same

charliescheer committed:
restores a test from a previous commit that had been removed by git

charliescheer committed:
Changed testNonEqualStringsCreateSameHash helper function private

charliescheer committed:
Simplified hash comparison test name for readability

charliescheer committed:
Simplified sample creation for testNonEqualStringsCreateSameHash

charliescheer committed:
Simplified test helper function testNonEqualStringsCreateSameHash

charliescheer commented:
>First we can make func testNonEqualStringsCreateSameHash(_ samples: [NSString]) private, because it's our internal helper method.

Complete

>testByEncodingTagAsHashAllowsUsToProperlyCompareStringsThatEvaluateAsNotEqualOtherwise to testNonEqualStringsCreateSameHash but without parameters.

much clearer, updated

>To solve this we can temporarily cast our strings to NSString: XCTAssertFalse((samples[count] as NSString).equal(to: samples[testCount]))

complete

>Another thing is that in original test we were comparing every sample to every sample, but I don't think it's needed. I think we can compare the first sample to all other samples and get the same coverage. If A == B and A == C, then B == C. This way we will need to walk the samples array only once.

That's a good point!  I have updated the helper function quite a bit to reflect this change.  I think it makes the whole test much easier to read

>p.s. you can use for loop instead of while + separate var: for i in 0..<samples.count
I had started with a for in loop, but for some reason found it cleaner to use the while when checking each property against each other, but after removing the b==c check I like the for in loop much better.

Let me know what you think!

eshurakov commented on SimplenoteTests/NSStringSimplenoteTests.swift:
```diff
@@ -24,21 +34,26 @@ class NSStringSimplenoteTests: XCTestCase {
+        ]
+        
+        testNonEqualStringsCreateSameHash(differentStringsOneHashSamples[0])
+        testNonEqualStringsCreateSameHash(differentStringsOneHashSamples[1])
```
Nitpick: this can be done in a loop, this way if we ever need to add more samples later, we can just add them to `differentStringsOneHashSamples`

eshurakov commented on SimplenoteTests/NSStringSimplenoteTests.swift:
```diff
@@ -24,21 +34,26 @@ class NSStringSimplenoteTests: XCTestCase {
+        let sampleToTest = samples[0]
+        
+        for i in 1..<samples.count {
+            XCTAssertNotEqual((sampleToTest as NSString), (samples[i] as NSString))
+            print("sample to test: \(sampleToTest) sample to comare: \(samples[i])")
+            XCTAssertEqual(sampleToTest.byEncodingAsTagHash, samples[i].byEncodingAsTagHash)
+            print("sample to test: \(sampleToTest.byEncodingAsTagHash) sample to comare: \(samples[i].byEncodingAsTagHash)")
```
Let's remove `print` :-)

charliescheer replied:
oops.... ;P

eshurakov reviewed (approved):
Left a few nitpick comments, but other than that looks great!

Let's :shipit:

charliescheer committed:
Set hash comparisons to test in a loop

charliescheer committed:
Remove print statement from test
